### DISCUSSION TITLE
How to add initial data to table in OnCreate method

### DISCUSSION THREAD

**Discussion by mytc** [simolus3/drift#1206] [Q&A] [answered]
So, I want to add initial data to one of my table once the app is installed.
This is what I did:
 ```
@override
  MigrationStrategy get migration {
    return MigrationStrategy(
        onCreate: (Migrator m) async {
          await m.createAll();
          return customStatement(
              "INSERT INTO users [name, wformat, hformat, tformat, bformat] VALUES ('Default User', 'kg', 'cm', '°C','mmol/L')");
        },
        onUpgrade: (Migrator m, int from, int to) async {},
        beforeOpen: (details) async {
          await customStatement('PRAGMA foreign_keys = ON');
        });
```
If I understand correctly the Custom Insert query statement need to run after `m.createAll();`, right? 
Problem is the `users` table is empty after installed the app. 

I am using `get_it` for the db provider. 
In my Doa, I have this:
`Future<List<User>> getAllUsers() => select(users).get();`
It return empty list when I call it at the UI. 
Anything wrong here?

**Comment by simolus3** [comment, marked as answer]
>If I understand correctly the Custom Insert query statement need to run after `m.createAll();`, right?

Yes, that is correct. Otherwise the table wouldn't exist. To me, it looks like you have a syntax error in your insert statement, AFAIK sqlite does not support `[` and `]` for column names which explains why no data is being added.

Note that in recent moor versions you can use the regular apis inside a migration callback, so maybe you prefer writing

```dart
onCreate: (Migrator m) async {
  await m.createAll();
  await into(users).insert(...);
},
```

Note that you may have to clear your app's data for this to work now since the db likely already exists. Still, it's a bit worrying that you didn't get an exception when opening the database for the first time. Did you overlook that maybe?

**Reply by mytc** [reply]
Yes, you are correct, it was the `[]` cause the problem.  Now it works.

But I have another question though. Is it the `onCreate` method run every time open the app? Because I noticed that every time I Run and Debug , I will see the  `Moor: Sent Create Table IF...`  statement in the Debug console. Is it intended to do so? If so the create table part is acceptable because there will be `IF NOT EXIST ` statement in the query.  But the Insert part will become the problem. I want 1 default record only in the table. However, so far all I can see only 1 record in the table, like the statement not affected the actual situation. 
Maybe it is due to emulator that cause the problem? It will do a fresh install when run in Debug mode?

**Reply by simolus3** [reply]
That definitely doesn't sound right, moor should only call `onCreate` once per database file. If you can reproduce this with a small sample I'd love to take a look at this, I don't think it's the emulator causing this.

**Reply by mytc** [reply]
I recheck many times, I found out sometime the console will stated that my emulator was insufficient of storage, then it will uninstall and reinstall again. I think this is the cause. In this case, I also wonder what make this development app took so many storage? The app I developing now still very early stage, which mean I only have very few packages and UI files added. The most complicated package that I installed probably is the moor . But maybe it was flutter problem rather then Moor itself. 
Anyway, thanks for prompt reply and help!
